Implicit namespaces support

gforcada opened the pull request:
Closes #160

Allow `zope.testrunner` to find tests suites for packages that follow the PEP 420 implicit namespaces.

Note that this works on test packages, like https://github.com/zopefoundation/megrok.strictrequire/pull/8

Test suite needs to be adjusted, but at least we have a reference implementation that _seems_ to be working for packages already using implicit namespaces.

To test it with `megrok.strictrequire`:

```shell
git clone git@github.com:zopefoundation/zope.testrunner
cd zope.testrunner
git checkout implicit-namespaces
git clone git@github.com:zopefoundation/megrok.strictrequire
cd megrok.strictrequire
git checkout pep440
cd ..
python3.11 -m venv venv
. venv/bin/activate
pip install -e megrok.strictrequire[test]
pip install -e .
zope-testrunner --test-path megrok.strictrequire/src
```

Most of the test failures of `zope.testrunner`'s own test suite is due to, probably, the test runner finding too much tests. I will dig into it later, but I wanted to post the PR as soon as it was minimally working ✨

gforcada commented:
Status update: all tests pass and the only two adjustments on the tests seem to be fine enough.

I could move those tests adjustments to a separate PR to validate that, indeed, they are not needed for this PR to pass, but stand on their own.

jensens reviewed (approved):
The only side effect is, it may take longer to discover tests, right? But I think this is fine. If someone complains, testrunner can be configured to search more specific.

dataflake commented:
> Test suite needs to be adjusted, but at least we have a reference implementation that _seems_ to be working for packages already using implicit namespaces.

What do you mean by "test suite needs to be adjusted"? Do you mean existing packages need their tests changed?

jensens commented:
> What do you mean by "test suite needs to be adjusted"? Do you mean existing packages need their tests changed?

Since it did not traverse into non-`__init__.py`-containing directories and now does, it may result in longer discovery runs. To avoid this we may want to configure the zope.testrunner to search more precisely using other `--path` and/or `--test-pattern` configurations.
OTOH I do not expect it to discover tests, which it had not found before - so, it won't break.

gforcada commented:
@dataflake sorry, that was an interim comment, when I first pushed this PR, I knew that the code was working, as I used the `megrok.staticrequire` distribution as a test sample, but I was not sure about `zope.testrunner`'s own test suite.

I was expecting more breakage, but turned out not to be 👍🏾 

Regarding test discovery, I see a problem, like we faced in `z3c.dependencychecker` when we added support for implicit namespaces (https://github.com/reinout/z3c.dependencychecker/pull/126) and the two follow up PRs to improve the time it takes to run: https://github.com/reinout/z3c.dependencychecker/pull/128

Basically, ignore folders with leading dots, `__pycache__` or `node_modules` for example.

dataflake commented:
> Basically, ignore folders with leading dots, `__pycache__` or `node_modules` for example.

It looks like the current code here only omits leading dot folders with the regular expression `identifier`:

https://github.com/zopefoundation/zope.testrunner/blob/a68a8669afee2e0248082a5ce40232a2249e5aaf/src/zope/testrunner/find.py#L28

This expression is used to filter out folders while walking the tree:

https://github.com/zopefoundation/zope.testrunner/blob/a68a8669afee2e0248082a5ce40232a2249e5aaf/src/zope/testrunner/find.py#L301

I suggest adding at least `__pycache__` to the undesirable folder names there.

davisagli commented:
I'm not sure what would be the simplest way to implement this, but it would be nice if the test discovery skipped any paths that are matched by `.gitignore` configuration.

icemac commented:
> I'm not sure what would be the simplest way to implement this, but it would be nice if the test discovery skipped any paths that are matched by `.gitignore` configuration.

There is https://github.com/thehanimo/py-gitignore under MIT license which seems to be capable of checking whether a path is ignored via `.gitignore`.

d-maurer commented:
Michael Howitz wrote at 2024-2-20 23:24 -0800:
>> I'm not sure what would be the simplest way to implement this, but it would be nice if the test discovery skipped any paths that are matched by `.gitignore` configuration.
>
>There is https://github.com/thehanimo/py-gitignore under MIT license which seems to be capable of checking whether a path is ignored via `.gitignore`.

Not sure whether depending on `.gitignore` is a good approach:
it would mean that test runs behave differently between
run from a git checkout (with .gitignore) and from e.g. an
installation (without ".gitignore")
(and maybe even a source distribution).

When I have understood this correctly, we run tests
not from a git checkout but from a wheel for some projects.

dataflake commented:
I agree with Dieter and I don't think there's much gain for the price of yet another dependency or manually written code that needs maintaining going forward. My suggestion still stands, expand the current exclusion of dot folders by adding `__pycache__` and get on with this.

davisagli commented:
@d-maurer Good point, I agree that's a compelling reason to not use `.gitignore`.

gforcada commented:
I adjusted the code just slightly to ignore well known folders we want to skip.

I added some minimal tests to ensure we know what we are accepting/rejecting right now.

🍀

jensens commented:
As an idea for a future iteration as a new PR (to get this one here merged soon): Get the `IGNORE_FOLDERS` configuration from a custom section in `pyproject.toml`?

gforcada commented:
@jensens yes, I thought about that as well, that one way or another users could provide their own list of ignored folders 👍🏾 

But yes, let's merge this, a new release, and see how much it breaks 🍀 😅

dataflake commented:
Release 6.4 is now published
